Hard computer player

gilroman opened the pull request.

Commit:
Add hard computer player

Commit:
Add game end message

pumpkincouture commented on src/tic_tac_toe_pair/ai.clj:
```diff
@@ -0,0 +1,79 @@
+  (:require [tic-tac-toe-pair.rules :refer [get-winning-token is-game-over?]]))
+
+(def win-score 100)
+(def loose-score -100)
```
This should be 'lose' instead of 'loose'

gilroman replied:
Thanks for catching this @pumpkincouture

pumpkincouture commented on src/tic_tac_toe_pair/console.clj:
```diff
@@ -107,8 +107,11 @@
+(defn build-congratulations-message [winning-token]
+  (str "Congratulations! " (keyword-to-token winning-token) " won the game!"))
+
+(defn build-loosing-message [winning-token]
```
Similarly, this should be 'losing'

ssunday commented on src/tic_tac_toe_pair/ai.clj:
```diff
@@ -0,0 +1,79 @@
+      (not (token (get-winning-token board)))loose-score))
+
+(defn get-board-score [board depth token is-maximizing] 
+(if is-maximizing
```
This should be indented in

gilroman replied:
Thanks!!!

ssunday commented on src/tic_tac_toe_pair/game.clj:
```diff
@@ -31,21 +41,23 @@
+      :else (build-loosing-message winner))))
+      
+(defmulti get-move :current-token)
+(defmethod get-move :player-1-token [game]
```
How would you refactor this so that either player could be computer or AI?

gilroman replied:
Thanks for reviewing my PR Sarah!!!
This is the way I would refactor the function. What do you think? 

1. Change the game map so that players can be given a type value of :computer or :human like this:

```
{:player-1 {:token :o :type :computer}
 :player-2 {:token :x :type :human}
 :board [nil nil nil nil nil nil nil nil]
 :current-player :player-1
 }
```

2. Refactor the function so that it calls the correct function depending on the player type.

```
(defn move [game]
  (let [{{current-player-type :type} (:current-player game)} game] 
    (cond 
      (= :computer current-player-type) (get-ai-move)
      (= :human current-player-type) (get-move-location))))
```

Having players be either ai or human is not part of this story, so I will save this work on a separate branch for future use.

ssunday replied:
That's like what I was thinking! I like giving the player both a marker and a type

gilroman replied:
Great! I definitely plan on refactoring and implementing this.

ssunday commented on src/tic_tac_toe_pair/rules.clj:
```diff
@@ -29,7 +29,7 @@
-(defn is-game-over? [game]
-  (if (or (is-full? (:board game)) (is-game-won? (:board game)))
+(defn is-game-over? [board]
+  (if (or (is-full? board) (is-game-won? board))
```
Could this be reduced to just being the boolean conditional? Do you need to return true/false like that?

gilroman replied:
Yes!!!! Great catch. I'll change it to:

`(defn is-game-over? [board]
  (or (is-full? board) (is-game-won? board)))`

ssunday replied:
That looks perfect

gilroman replied:
Sarah, thank you for your time looking at this. Do you approve this PR? 
I've been guided to make sure to get explicit approval before merging it.

ssunday replied:
@gilroman Oh my gosh, I meant to hit approve but then I got lost in other stuff. Sorry!!!!!!

gilroman replied:
@ssunday No worries at all!!! It was great getting your feedback.

Commit:
Address feedback

- Address feedback in regards to typo of word lose
- Address feedback in regards to use of conditional in is-game-over? function
